### PR TITLE
[Core] Update actor cancelation API documentation & API docstring

### LINKED ISSUE #34961: [core] cancelling actor tasks
We have multiple feature requests in Ray Serve to cancel tasks / handle HTTP timeouts

- https://github.com/ray-project/ray/issues/22450
- https://github.com/ray-project/ray/issues/34682
- https://github.com/ray-project/ray/issues/34841

It would be simple to implement these if it were possible to `ray.cancel` an object ref to an actor task in an asyncio actor. Ideally this would use the `asyncio.cancel` API so we can handle it gracefully.

### PR DESCRIPTION
**Pull request by rkooo567**
## Why are these changes needed?

This PR updates the docstring of `ray.cancel` and a documentation to update Actor cancelation API.

## Related issue number

Closes https://github.com/ray-project/ray/issues/34961

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by angelinalg** [comment]
@rkooo567 Apologies for the barrage of suggestions. They are mostly for capitalization of `Actors` and `Tasks`, which I believe we do when we are talking about Ray Actors and Tasks. Please correct if I misinterpreted. The remaining suggestions are copy edits. Thanks for updating the documentation!!

File: doc/source/ray-core/actors.rst
Comment (edoakes) [inline doc/source/ray-core/actors.rst:383]:
```suggestion
Cancelling actor tasks
```

File: doc/source/ray-core/actors.rst
Comment (edoakes) [inline doc/source/ray-core/actors.rst:386]:
```suggestion
Actor tasks can be canceled by calling :func:`ray.cancel() <ray.cancel>` on the returned `ObjectRef`.
```

File: doc/source/ray-core/actors.rst
Comment (edoakes) [inline doc/source/ray-core/actors.rst:410]:
```suggestion
For tasks classified as `async actors <_async-actors>`, Ray seeks to cancel the associated `asyncio.Task`. 
```
coroutines cannot be cancelled, it's task-specific

File: doc/source/ray-core/actors.rst
Comment (edoakes) [inline doc/source/ray-core/actors.rst:414]: please add a note that if you don't `await`, the task won't be interrupted

File: doc/source/ray-core/actors.rst
Comment (edoakes) [inline doc/source/ray-core/actors.rst:407]: future work: we should really just add a `ray.is_cancelled()` method that folks can poll (I think you mentioned it in the past)

Comment (rkooo567) [reply doc/source/ray-core/actors.rst:407]: Let me create an issue

**Comment by rkooo567** [comment]
All the comments are addressed

**Comment by rkooo567** [comment]
We won't merge it until API is approved

**Review by angelinalg — APPROVED** [review]

**Comment by rkooo567** [comment]
All comments are addressed. @edoakes can you approve it if it looks okay?

**Review by edoakes — APPROVED** [review]

File: python/ray/_private/worker.py
Comment (pcmoritz) [inline python/ray/_private/worker.py:2877]: This should be `ValueError` instead of `TypeError`. Since we just introduced this with https://github.com/ray-project/ray/commit/03ae7790fb8c11a40f88c3186baf6556a2eaecdd, I believe we can still change it before the 2.7 release, right?

Comment (pcmoritz) [reply python/ray/_private/worker.py:2877]: Also in my testing I noticed that this is not raising an error if the task already finished. This is pretty confusing behavior I think:
```
In [3]: @ray.remote
   ...: class Actor:
   ...:     def f(self):
   ...:         print("hi")
   ...: 

In [4]: a = Actor.remote()

In [5]: x = a.f.remote()

(Actor pid=6797) hi
In [6]: ray.cancel(x, force=True)
```

File: python/ray/_private/worker.py
Comment (pcmoritz) [inline python/ray/_private/worker.py:2857]: These different behaviors are strange, but I assume we can't really change it any more without API changes, right (i.e. this is already the behavior or Ray 2.6)?

**Review by pcmoritz — APPROVED** [review]
Please address the comments and then this should be merged for 2.7 :)

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2877]: Let me try fixing it now... IIRC, this didn't work (when a task is finished, we don't have task spec anymore, and there's no way to obtain if object ref is from task or actor rn. I tried object_ref.actor_id().IsNil(), but it looks like this was buggy). Do you think this PR is mergeable when this issue is fixed, or is it something we can fix as a follow up?

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py:2857]: Oh actually I fixed this in the master because ray.cancel never specifies this behavior in the docstring. If you read the docstring, it says

```
    Calling ray.get on a canceled task will raise a TaskCancelledError or a
    WorkerCrashedError if ``force=True``.
```

So I considered as a bug fix, not API change. https://github.com/ray-project/ray/pull/38973. (I also brought this up to api review channel, https://anyscaleteam.slack.com/archives/CTSRWH3DJ/p1692224223820259)

File: python/ray/_private/worker.py
Comment (pcmoritz) [reply python/ray/_private/worker.py:2877]: Yeah, we can fix this as a follow up. But the `ValueError` versus `TypeError` needs to be fixed before merging (my first comment) because it is irreversible without an API change.

File: python/ray/_private/worker.py
Comment (jjyao) [reply python/ray/_private/worker.py:2857]: We then should update the doc to reflect the bug fix?

TaskCancelledError if the Task has been scheduled and interrupted,

File: python/ray/_private/worker.py
Comment (rkooo567) [reply python/ray/_private/worker.py]: Btw, thanks for catching this! (little embarassing, but I learned ValueError means the invalid input first time...)

**Comment by rkooo567** [comment]
@pcmoritz can you check the new changes? https://github.com/ray-project/ray/pull/38776/commits/8b17f8f254019698b99dd0de5348ef42ee8847a4

I will merge once tests pass!

File: doc/source/ray-core/actors.rst
Comment (rkooo567) [reply doc/source/ray-core/actors.rst]: https://github.com/ray-project/ray/issues/39152

**Pull request merged**
